source/slang/slang-ir-byte-address-legalize.cpp

```diff
@@ -0,0 +1,924 @@
+// slang-ir-byte-address-legalize.cpp
+#include "slang-ir-byte-address-legalize.h"
+
+// This file implements an IR pass that translates load/store operations
+// on byte-address buffers to be legal for a chosen target.
+//
+// Currently this pass only applies to the operations generated for
+// the generic `*ByteAddressBuffer.Load<T>` and `.Store<T>` operations,
+// and not the non-generic versions that traffic in `uint` (e.g.,
+// `Load2` or `Store3`).
+
+#include "slang-ir-insts.h"
+#include "slang-ir-layout.h"
+
+namespace Slang
+{
+
+// As is typical for IR passes in Slang, we will encapsulate the state
+// while we process the code in a context type.
+//
+struct ByteAddressBufferLegalizationContext
+{
+    // We need access to the original session, as well as the options
+    // that control what constructs we legalize, and how.
+    //
+    Session* m_session = nullptr;
+    ByteAddressBufferLegalizationOptions m_options;
+
+    // We will also use a central IR builder when generating new
+    // code as part of legalization (rather than create/destroy
+    // IR builders on the fly).
+    //
+    SharedIRBuilder m_sharedBuilder;
+    IRBuilder m_builder;
+
+    // Everything starts with a request to process a module,
+    // which delegates to the central recrusive walk of the IR.
+    //
+    void processModule(IRModule* module)
+    {
+        m_sharedBuilder.session = m_session;
+        m_sharedBuilder.module = module;
+
+        m_builder.sharedBuilder = &m_sharedBuilder;
+
+        processInstRec(module->getModuleInst());
+    }
+
+    // We recursively walk the entire IR structure (except
```

Comment:
This isn't something for this PR, but based on something we discussed earlier - the appeal of traversing IR with recursive traversals of the entire module tree. It's very appealing because it's so simple.

That an advantage to handling off traversal to some kind of 'traversal manager' even with the homogeneity that IR gives us, is that it can achieve two things (depending on the type of transformation).

1) Could allow a single traversal to do multiple (typically simple) transformations
2) Can use knowledge about the structure to speed traversal - if we are looking for a node type that we know can't exist other node types, we can just skip that
3) Could hold a cache that could speed other kinds of searches/transitions

How much this helps in practice is arguable, and you could imagine you might want more than one style of traversal manager, and the baked in knowledge could be quite esoteric.

Reply:
I agree that it makes some sense to offload the main recursive walk to subroutine, whether or not we decide to make the kinds of improvements you are talking about. Repeating that code over and over is a waste of time and energy, really.

For my part I'd be interested to see such a change comes a part of codifying the notion of a "pass" in the IR, so that things like optimization first construct an explicit list of `Pass` objects to apply, and then a central execution engine goes through and applies those passes.

The execution engine could then own things like:

* Running validation before/after passes
* Dumping of IR before/after passes (including controls for picking where in the process to dump)
* Ganging together passes that can share the recursive walk over the tree (as you describe)
* Parallelizing passes over functions, where that is possible
* Computing and then re-using side-band structures like the dominator tree across passes when intermediate passes do not invalidate it

All of that is squarely in the "would be nice" pile rather than the "must have" pile right now.